### PR TITLE
alloc profile: move into Profile package

### PR DESCRIPTION
**Pull request by vilterp**
Stacked on #42768

As of this PR the API is

```julia
using Profile
Profile.Allocs.@profile skip_every=1 begin
  # ...
end
```

Might be nice to re-export it so it's

```julia
using Profile
Profile.@profile_allocs skip_every=1 begin
  # ...
end
```

Not sure the most idiomatic way to do that. Declare a macro in Profile and just have it call the one in `Profile.Allocs`?

### CODE REVIEW

**Comment by NHDaly** [comment]
Yeah, I think this is a nice change. 👍 

Adding @vchuravy, @IanButterworth for your opinions on the profiling API.

Gosh, i'm really bad at naming bikeshedding.. I see merit to both options you've presented. You could also call it `Profile.@allocs_profile`, as yet another option.

I don't have much preference between those three.

And yeah, for implementing a macro like you describe, you can forward the macro call, yeah. Something like this, in the `Profile` module:
```julia
macro profile_allocs(args...)
    return esc(:($Allocs.@profile $(args...)))
end
```

File: stdlib/Profile/src/Profile.jl [hunk 1]
```diff
@@ -1139,4 +1139,6 @@ function warning_empty(;summary = false)
     end
 end
 
+include("Allocs.jl")
+
```
Comment (NHDaly) [inline stdlib/Profile/src/Profile.jl:1143]: Does this need to go at the end? I think usually the standard is to put `include`s at the top of the module. If it needs to use names from the parent module, it'd probably be preferable to move the whole of the Profile.jl file into another file (call it `cpu_profile.jl`?) and include that before including this?

But ideally you can just put this at the top (after the imports/usings) and it'll just work? 🤞

**Review by NHDaly — APPROVED** [review]
LGTM 👍 👍

**Comment by IanButterworth** [comment]
How much overhead does the recording add?
If it's small, perhaps this could just be added to `Profile.@profile`?

I assume no report prints immediately after the macro finishes recording? If so, the user can then do

```
using Profile

@profile foo()
```
Then decide whether to look at `Profile.print()` or the allocation report?

**Comment by IanButterworth** [comment]
Or just merge the two reports??

**Comment by NHDaly** [comment]
That's an interesting idea..

I think there are maybe enough drawbacks not to support that option though:
- The main drawback is that we decided not to go through the pains of making users preallocate a buffer to store the results like `Profile.@profile` does, and to instead use a dynamically sized C++ vector, but this means that the time spent resizing the vector would incorrectly show up in your CPU profile... Which probably isn't great?
- Another is that this has another argument to be configured (`skip_every=n`) so we'd still need to add a way to config that, and the API gets messy there.. If we mixed it into the cpu profiling, you'd have to prefix it with `allocs_`, and might have something like `Profile.init(delay=0.01, allocs_skip_every=n)`, which is a bit messy?
- Another is that currently we _are_ returning the profile results from the `@profile_allocs` macro, though maybe we shouldn't be... 🤔 (Since you might still want to get the results of the expression being profiled, i think?)
- Finally, yeah, if you don't set a high `skip_every=n` number, this actually does have a lot of overhead, yeah.. :/ We could maybe reduce that, but it's probably still best to let users decide which they want?

As another data point, in the Golang pprof profiler, it seems to have separate top-level invocations:
- https://go.dev/doc/diagnostics#profiling

But it _is_ an interesting thought, and we could change it if we decide to!

**Comment by NHDaly** [comment]
(also, thanks for the quick reply, Ian! ❤️)

**Comment by vilterp** [comment]
Yeah, given that the overhead is nontrivial, I think we're gonna keep them separate for now. They should be ok to run at the same time though if you want to; may be nice to offer an API like `@profile_both` or something…

**Pull request merged**